lib/luaxtable.c

```diff
@@ -43,37 +47,123 @@ typedef struct luaxtable_s {
 		struct xt_match match;
 		struct xt_target target;
 	};
+	int fallback;
 	luaxtable_type_t type;
 } luaxtable_t;
 
-static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
-{
-	return true;
-}
+static lunatik_object_t *xtable_targethooks = NULL;
+static lunatik_object_t *xtable_matchhooks = NULL;
 
-static int luaxtable_match_check(const struct xt_mtchk_param *par)
+static inline void luaxtable_newdata(lua_State *L, void *ptr, size_t size, bool sleep)
 {
-	return 0;
+	lunatik_require(L, data);
+	lunatik_object_t *data = lunatik_checknull(L, luadata_new(ptr, size, sleep));
+	lunatik_cloneobject(L, data);
 }
 
-static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+#define luaxtable_getrcu(xtable, hook, key, keylen)			\
+do {									\
+	lunatik_object_t *obj = luarcu_gettable(xtable_##hook##hooks, key, keylen);	\
+	if (obj == NULL) {						\
+		pr_err("could not find table for %s\n", key);		\
+		return -1;						\
+	}								\
+	xtable = (luaxtable_t *)obj->private;				\
+} while(0)
+
+#define luaxtable_setrcu(object, hook, key, keylen)			\
+do {									\
+	if (luarcu_settable(xtable_##hook##hooks, key, keylen, object) != 0) {	\
+		luaL_error(L, "unable to set table");			\
+		return -1;						\
+	}								\
+} while(0)
+
+
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, struct sk_buff *skb, const char *op, int nret)
 {
+	int base = lua_gettop(L);
+	int ret = -ENXIO;
+	int nargs = 0;
+
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("could not find ops table\n");
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	if (skb != NULL ){
+		luaxtable_newdata(L, skb, sizeof(struct sk_buff), false);
+		nargs = 1;
+	}
+
+	if (lua_pcall(L, nargs, nret, 0) != LUA_OK) {
+		pr_err("%s error: %s\n", op, lua_tostring(L, -1));
+		ret = -ECANCELED;
+		goto err;
+	}
+
+	return nret > 0 ? luaL_optinteger(L, -1, 0) : 0;
```

Comment:
```suggestion
	return luaL_optinteger(L, -1, 0);
```

I think it should be enough, doesn't it? take a look at the manual..